src/java/org/apache/cassandra/db/compaction/unified/Controller.java

```diff
@@ -202,36 +249,180 @@ public int getThreshold(int index) {
+                                 localDensity / minSize,
+                                 FBUtilities.prettyPrintBinary(minSize, "B", " "));
+
+                return shards;
+            }
+        }
+
+        if (sstableGrowthModifier == 1)
+        {
+            shards = baseShardCount;
+            logger.debug("Shard count {} for density {} in fixed shards mode",
+                         shards,
+                         FBUtilities.prettyPrintBinary(localDensity, "B", " "));
+            return shards;
+        }
+        else if (sstableGrowthModifier == 0)
+        {
+            // How many we would have to aim for the target size. Divided by the base shard count, so that we can ensure
+            // the result is a multiple of it by multiplying back below. Adjusted by sqrt(0.5) to calculate the split
+            // points needed for the minimum size.
+            double count = localDensity / (targetSSTableSize * INVERSE_SQRT_2 * baseShardCount);
+            if (count > MAX_SHARD_SPLIT)
+                count = MAX_SHARD_SPLIT;
+            assert !(count < 0);    // Must be positive, 0 or NaN, which should translate to baseShardCount
+
+            // Make it a power of two multiple of the base count so that split points for lower levels remain split points for higher.
+            // The conversion to int and highestOneBit round down, for which we compensate by using the sqrt(0.5) multiplier
+            // already applied in the count.
+            // Setting the bottom bit to 1 ensures the result is at least baseShardCount.
+            shards = baseShardCount * Integer.highestOneBit((int) count | 1);
+
+            if (logger.isDebugEnabled())
+                logger.debug("Shard count {} for density {}, {} times target {}",
+                             shards,
+                             FBUtilities.prettyPrintBinary(localDensity, "B", " "),
+                             localDensity / targetSSTableSize,
+                             FBUtilities.prettyPrintBinary(targetSSTableSize, "B", " "));
+            return shards;
+        }
+        else
+        {
+            // How many we would have to aim for the target size. Divided by the base shard count, so that we can ensure
+            // the result is a multiple of it by multiplying back below.
+            double count = localDensity / (targetSSTableSize * baseShardCount);
+            // Take a logarithm of the count (in base 2), and adjust it by the given growth modifier.
+            // Adjust by 0.5 to round the exponent so that the result falls between targetSSTableSize * sqrt(0.5) and
+            // targetSSTableSize * sqrt(2). Finally, make sure the exponent is at least 0 and not greater than the
+            // fixed maximum.
+            // Note: This code also works correctly for the special cases of sstableGrowthModifier == 0 and 1,
+            //       but the above code avoids the floating point arithmetic for these common cases.
```

Comment:
Nit: We can add "imprecise" before "floating point arithmetic" as the loss of precision is the main reason to avoid this path.